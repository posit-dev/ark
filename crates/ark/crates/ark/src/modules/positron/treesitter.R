#
# treesitter.R
#
# Copyright (C) 2023-2024 Posit Software, PBC. All rights reserved.
#
#

#' @export
.ps.treesitter.node.text <- function(node, contents) {
    .ps.Call("ps_treesitter_node_text", node, contents)
}

#' @export
.ps.treesitter.node.kind <- function(node) {
    .ps.Call("ps_treesitter_node_kind", node)
}
