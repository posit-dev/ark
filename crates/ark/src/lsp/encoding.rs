//
// encoding.rs
//
// Copyright (C) 2024 Posit Software, PBC. All rights reserved.
//
//

use ropey::Rope;
use tower_lsp::lsp_types::Position;
use tree_sitter::Point;

/// `PositionEncodingKind` describes the encoding used for the `Position` `character`
/// column offset field. The `Position` `line` field is encoding agnostic, but the
/// `character` field specifies the number of characters offset from the beginning of
/// the line, and the "character" size is dependent on the encoding. The LSP specification
/// states:
///
/// - UTF8: Character offsets count UTF-8 code units (e.g. bytes).
/// - UTF16: Character offsets count UTF-16 code units (default).
/// - UTF32: Character offsets count UTF-32 code units (these are the same as Unicode
///   codepoints, so this `PositionEncodingKind` may also be used for an encoding-agnostic
///   representation of character offsets.)
///
/// The `vscode-languageclient` library that Positron uses on the frontend to create the
/// `Client` side of the LSP currently ONLY supports `UTF16`, and will error on anything
/// else. Their reasoning is that it is easier for the server (ark) to do the re-encoding,
/// since we are tracking the full document state. Track support for UTF-8 here:
/// https://github.com/microsoft/vscode-languageserver-node/issues/1224
///
/// The other interesting part of this is that `TextDocumentContentChangeEvent`s that
/// come through the `did_change()` event and the `TextDocumentItem` that comes through
/// the `did_open()` event encode the `text` of the change/document in UTF-8, even though
/// the `Range` (in the case of `did_change()`) that tells you where to apply the change
/// uses UTF-16, so that's cool. UTF-8 `text` is forced to come through due to how the
/// LSP specification uses jsonrpc, where the content fields must be 'utf-8' encoded:
/// https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#contentPart
/// This at least means we have a guarantee that the document itself and any updates to
/// it will be encoded in UTF-8, even if the Positions are UTF-16.
///
/// So we need a way to convert the UTF-16 `Position`s to UTF-8 `tree_sitter::Point`s and
/// back. This requires the document itself, and is what the helpers in this file implement.
pub fn get_position_encoding_kind() -> tower_lsp::lsp_types::PositionEncodingKind {
    tower_lsp::lsp_types::PositionEncodingKind::UTF16
}

pub fn convert_tree_sitter_range_to_lsp_range(
    x: &Rope,
    range: tree_sitter::Range,
) -> tower_lsp::lsp_types::Range {
    let start = convert_point_to_position(x, range.start_point);
    let end = convert_point_to_position(x, range.end_point);
    tower_lsp::lsp_types::Range::new(start, end)
}

pub fn convert_position_to_point(x: &Rope, position: Position) -> Point {
    let line = position.line as usize;
    let character = position.character as usize;

    let character = with_line(x, line, character, convert_character_from_utf16_to_utf8);

    Point::new(line, character)
}

pub fn convert_point_to_position(x: &Rope, point: Point) -> Position {
    let line = point.row;
    let character = point.column;

    let character = with_line(x, line, character, convert_character_from_utf8_to_utf16);

    let line = line as u32;
    let character = character as u32;

    Position::new(line, character)
}

fn with_line<F>(x: &Rope, line: usize, character: usize, f: F) -> usize
where
    F: FnOnce(&str, usize) -> usize,
{
    let Some(x) = x.get_line(line) else {
        let n = x.len_lines();
        let x = x.to_string();
        let line = line + 1;
        // Forcing a full capture so we can learn the situations in which this occurs
        let trace = std::backtrace::Backtrace::force_capture();
        log::error!(
            "Requesting line {line} but only {n} lines exist.\n\nDocument:\n{x}\n\nBacktrace:\n{trace}"
        );
        return 0;
    };

    // If the line is fully contained in a single chunk (likely is), use free conversion to `&str`
    if let Some(x) = x.as_str() {
        return f(x, character);
    }

    // Otherwise, use ever so slightly more expensive String materialization of the
    // line spread across chunks
    let x = x.to_string();
    let x = x.as_str();

    f(x, character)
}

/// Converts a character offset into a particular line from UTF-16 to UTF-8
fn convert_character_from_utf16_to_utf8(x: &str, character: usize) -> usize {
    if x.is_ascii() {
        // Fast pass
        return character;
    }

    // Initial check, since loop would skip this case
    if character == 0 {
        return character;
    }

    let mut n = 0;

    // For each `u32` sized `char`, figure out the equivalent size in UTF-16
    // world of that `char`. Once we hit the requested number of `character`s,
    // that means we have indexed into `x` to the correct position, at which
    // point we can take the current bytes based `pos` that marks the start of
    // this `char`, and add on its UTF-8 based size to return an adjusted column
    // offset. We use `==` because I'm fairly certain they should always align
    // exactly, and it would be good to log if that isn't the case.
    for (pos, char) in x.char_indices() {
        n += char.len_utf16();

        if n == character {
            return pos + char.len_utf8();
        }
    }

    log::error!("Failed to locate UTF-16 offset of {character}. Line: '{x}'.");
    return 0;
}

/// Converts a character offset into a particular line from UTF-8 to UTF-16
fn convert_character_from_utf8_to_utf16(x: &str, character: usize) -> usize {
    if x.is_ascii() {
        // Fast pass
        return character;
    }

    // The UTF-8 -> UTF-16 case is slightly simpler. We just slice into `x`
    // using our existing UTF-8 offset, reencode the slice as a UTF-16 based
    // iterator, and count up the pieces.
    match x.get(..character) {
        Some(x) => x.encode_utf16().count(),
        None => {
            let n = x.len();
            log::error!(
                "Tried to take UTF-8 character {character}, but only {n} characters exist. Line: '{x}'."
            );
            0
        },
    }
}
