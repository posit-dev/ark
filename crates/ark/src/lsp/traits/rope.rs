// 
// rope.rs
// 
// Copyright (C) 2022 by RStudio, PBC
// 
// 

use ropey::Rope;
use tower_lsp::lsp_types::Position;
use tree_sitter::Point;

pub(crate) trait RopeExt {
    fn point_to_byte(&self, point: Point) -> usize;
    fn position_to_byte(&self, position: Position) -> usize;
}

impl RopeExt for Rope {

    // TODO: We likely need to translate column positions into byte positions,
    // to properly handle multibyte unicode characters.
    fn point_to_byte(&self, point: Point) -> usize {
        self.line_to_byte(point.row) + point.column
    }

    fn position_to_byte(&self, position: Position) -> usize {
        self.line_to_byte(position.line as usize) + position.character as usize
    }

}
