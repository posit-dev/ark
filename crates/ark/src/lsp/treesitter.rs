//
// treesitter.rs
//
// Copyright (C) 2023 Posit Software, PBC. All rights reserved.
//
//

use harp::external_ptr::ExternalPointer;
use harp::object::RObject;
use libR_sys::SEXP;
use tree_sitter::Node;

#[harp::register]
pub unsafe extern "C" fn ps_treesitter_node_text(node_ptr: SEXP, source_ptr: SEXP) -> SEXP {
    let node = ExternalPointer::<Node>::reference(node_ptr);
    let source = ExternalPointer::<&str>::reference(source_ptr);

    let text = node.utf8_text(source.as_bytes()).unwrap_or("");
    *RObject::from(text)
}

#[harp::register]
pub unsafe extern "C" fn ps_treesitter_node_kind(node_ptr: SEXP) -> SEXP {
    let node = ExternalPointer::<Node>::reference(node_ptr);

    *RObject::from(node.kind())
}
