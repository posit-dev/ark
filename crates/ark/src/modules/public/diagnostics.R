#
# completions.R
#
# Copyright (C) 2022 Posit Software, PBC. All rights reserved.
#
#

.ps.diagnostics.diagnostic <- function(kind = "skip", node = NULL, message = NULL) {
    list(kind, node, message)
}

.ps.diagnostics.custom.library <- function(call, contents, fun = base::library) {
    # TODO: it could be interesting to have a diagnostic
    #       when this will fail on wrong argument, e.g.
    #       library(pkg = ggplot2)
    #               ^^^ : unused argument (pkg)
    matched_call <- match.call(fun, call)

    # here we get a call where arguments are named, e.g.
    # library(package = <x>) where <x> is a list of 2 things:
    # - a 0-based integer position for this argument. This is not
    #   currently used
    # - an external pointer to a treesitter Node, which can be queried
    #   with .ps.treesitter.node.text() and .ps.treesitter.node.kind()
    #
    # We might simplify and only pass around the external pointer if
    # we realize the position isn't useful.

    # identify if character.only was set, so that we can
    # adapt the diagnostic appropriately
    is_character_only <- function(matched_call, contents) {
        character_only <- matched_call[["character.only"]]

        if (is.null(character_only)) {
            FALSE
        } else {
            ptr <- character_only[[2L]]
            text <- .ps.treesitter.node.text(ptr, contents)
            !identical(text, "FALSE")
        }
    }

    # deal with arguments `package` and `help` which use
    # non standard evaluation, e.g. library(ggplot2)
    diagnostic_package <- function(arg, contents, character_only) {
        index <- arg[[1L]]
        node <- arg[[2L]]

        kind <- .ps.treesitter.node.kind(node)

        # library(foo, character.only = TRUE)
        if (kind == "identifier" && character_only) {
            return(.ps.diagnostics.diagnostic("default", node))
        }

        if (kind %in% c("string", "identifier")) {
            # library("foo") or library(foo)
            pkg <- .ps.treesitter.node.text(node, contents)

            if (kind == "string") {
                pkg <- gsub("^(['\"])(.*)\\1$", "\\2", pkg)
            }

            # The package is installed, just skip the diagnostic
            if (pkg %in% base::.packages(all.available = TRUE)) {
                return(.ps.diagnostics.diagnostic("skip"))
            }

            msg <- sprintf("Package '%s' is not installed", pkg)
            return(.ps.diagnostics.diagnostic("simple", node, message = msg))
        }

        .ps.diagnostics.diagnostic("default", node)
    }

    # Before scanning all arguments, we need to check if
    # character.only is set, so that we can adapt how the
    # package and help arguments are handled
    character_only <- is_character_only(matched_call, contents)

    # Scan the given arguments and make diagnostics for each
    n <- length(matched_call)
    out <- vector("list", n - 1L)
    names <- names(matched_call)
    for (i in seq_len(n - 1L)) {
        arg <- matched_call[[i + 1L]]
        name <- names[[i + 1L]]

        diagnostic <- if (name %in% c("package", "help")) {
            diagnostic_package(arg, contents, character_only)
        } else {
            .ps.diagnostics.diagnostic("default", node = arg[[2L]])
        }

        out[[i]] <- diagnostic
    }

    out
}

.ps.diagnostics.custom.require <- function(call, contents, fun = base::require) {
    .ps.diagnostics.custom.library(call, contents, fun)
}
