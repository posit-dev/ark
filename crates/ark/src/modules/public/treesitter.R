
#
# treesitter.R
#
# Copyright (C) 2023 Posit Software, PBC. All rights reserved.
#
#

.ps.treesitter.node.text <- function(node, contents) {
    .ps.Call("ps_treesitter_node_text", node, contents)
}

.ps.treesitter.node.kind <- function(node) {
    .ps.Call("ps_treesitter_node_kind", node)
}
