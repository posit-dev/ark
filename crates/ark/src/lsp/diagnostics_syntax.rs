//
// diagnostics_syntax.rs
//
// Copyright (C) 2024 Posit Software, PBC. All rights reserved.
//
//

use tower_lsp::lsp_types::Diagnostic;
use tree_sitter::Node;
use tree_sitter::Range;

use crate::lsp::diagnostics::DiagnosticContext;
use crate::lsp::encoding::convert_tree_sitter_range_to_lsp_range;
use crate::lsp::traits::rope::RopeExt;
use crate::treesitter::node_has_error_or_missing;
use crate::treesitter::NodeType;
use crate::treesitter::NodeTypeExt;

pub(crate) fn syntax_diagnostics(
    root: Node,
    context: &DiagnosticContext,
) -> anyhow::Result<Vec<Diagnostic>> {
    let mut diagnostics = Vec::new();

    recurse(root, context, &mut diagnostics)?;

    Ok(diagnostics)
}

// When we hit an `ERROR` node, i.e. a syntax error, it often has its own children
// which can also be `ERROR`s. The goal is to target the deepest (most precise) `ERROR`
// nodes and only report syntax errors for those. We accomplish this by recursing
// into children first and bailing if we find any children that we reported an error for.
fn recurse(
    node: Node,
    context: &DiagnosticContext,
    diagnostics: &mut Vec<Diagnostic>,
) -> anyhow::Result<bool> {
    if !node_has_error_or_missing(&node) {
        // Stop recursion if this branch of the tree doesn't have issues
        return Ok(false);
    }

    // Always look for contextual `MISSING` issues based on the current node type
    diagnose_missing(node, context, diagnostics)?;

    let mut any_errors = recurse_children(node, context, diagnostics)?;

    // Report an error when:
    // - No children were `ERROR`s
    // - We are an `ERROR`
    if !any_errors && node.is_error() {
        diagnostics.push(syntax_diagnostic(node, context)?);
        any_errors = true;
    }

    Ok(any_errors)
}

fn recurse_children(
    node: Node,
    context: &DiagnosticContext,
    diagnostics: &mut Vec<Diagnostic>,
) -> anyhow::Result<bool> {
    let mut any_errors = false;
    let mut cursor = node.walk();

    for child in node.children(&mut cursor) {
        any_errors |= recurse(child, context, diagnostics)?;
    }

    Ok(any_errors)
}

fn syntax_diagnostic(node: Node, context: &DiagnosticContext) -> anyhow::Result<Diagnostic> {
    if let Some(diagnostic) = syntax_diagnostic_missing_open(node, context)? {
        return Ok(diagnostic);
    }

    Ok(syntax_diagnostic_default(node, context))
}

// Use a heuristic that if we see a syntax error and it just contains a `)`, `}`, or `]`,
// then it is probably a case of missing a matching open token.
fn syntax_diagnostic_missing_open(
    node: Node,
    context: &DiagnosticContext,
) -> anyhow::Result<Option<Diagnostic>> {
    let text = context.contents.node_slice(&node)?;

    let open_token = if text == ")" {
        "("
    } else if text == "}" {
        "{"
    } else if text == "]" {
        "["
    } else {
        // Not an unmatched closing token
        return Ok(None);
    };

    let range = node.range();

    Ok(Some(new_missing_open_diagnostic(
        open_token, range, context,
    )))
}

fn syntax_diagnostic_default(node: Node, context: &DiagnosticContext) -> Diagnostic {
    let range = node.range();
    let row_span = range.end_point.row - range.start_point.row;

    if row_span >= 20 {
        return syntax_diagnostic_truncated_default(range, context);
    }

    // The most common case, a localized syntax error that doesn't span too many rows
    let message = String::from("Syntax error");
    new_syntax_diagnostic(message, range, &context)
}

// If the syntax error spans more than 20 rows, just target the starting position
// to avoid overwhelming the user.
fn syntax_diagnostic_truncated_default(range: Range, context: &DiagnosticContext) -> Diagnostic {
    // In theory this is an empty range, as they are constructed like `[ )`, but it
    // seems to work for the purpose of diagnostics, and getting the correct
    // coordinates exactly right seems challenging.
    let start_range = Range {
        start_byte: range.start_byte,
        start_point: range.start_point,
        end_byte: range.start_byte,
        end_point: range.start_point,
    };

    // `+1` because it is user facing and editor UI is 1-indexed
    let end_row = range.end_point.row + 1;
    let message = format!("Syntax error. Starts here and ends on line {end_row}.");

    new_syntax_diagnostic(message, start_range, &context)
}

fn diagnose_missing(
    node: Node,
    context: &DiagnosticContext,
    diagnostics: &mut Vec<Diagnostic>,
) -> anyhow::Result<()> {
    match node.node_type() {
        NodeType::Parameters => diagnose_missing_parameters(node, context, diagnostics),
        NodeType::BracedExpression => {
            diagnose_missing_braced_expression(node, context, diagnostics)
        },
        NodeType::ParenthesizedExpression => {
            diagnose_missing_parenthesized_expression(node, context, diagnostics)
        },
        NodeType::Call => diagnose_missing_call(node, context, diagnostics),
        NodeType::Subset => diagnose_missing_subset(node, context, diagnostics),
        NodeType::Subset2 => diagnose_missing_subset2(node, context, diagnostics),
        NodeType::BinaryOperator(_) => diagnose_missing_binary_operator(node, context, diagnostics),
        _ => Ok(()),
    }
}

fn diagnose_missing_parameters(
    node: Node,
    context: &DiagnosticContext,
    diagnostics: &mut Vec<Diagnostic>,
) -> anyhow::Result<()> {
    diagnose_missing_close(node, context, diagnostics, ")")
}

fn diagnose_missing_braced_expression(
    node: Node,
    context: &DiagnosticContext,
    diagnostics: &mut Vec<Diagnostic>,
) -> anyhow::Result<()> {
    diagnose_missing_close(node, context, diagnostics, "}")
}

fn diagnose_missing_parenthesized_expression(
    node: Node,
    context: &DiagnosticContext,
    diagnostics: &mut Vec<Diagnostic>,
) -> anyhow::Result<()> {
    diagnose_missing_close(node, context, diagnostics, ")")
}

fn diagnose_missing_call(
    node: Node,
    context: &DiagnosticContext,
    diagnostics: &mut Vec<Diagnostic>,
) -> anyhow::Result<()> {
    diagnose_missing_call_like(node, context, diagnostics, ")")
}

fn diagnose_missing_subset(
    node: Node,
    context: &DiagnosticContext,
    diagnostics: &mut Vec<Diagnostic>,
) -> anyhow::Result<()> {
    diagnose_missing_call_like(node, context, diagnostics, "]")
}

fn diagnose_missing_subset2(
    node: Node,
    context: &DiagnosticContext,
    diagnostics: &mut Vec<Diagnostic>,
) -> anyhow::Result<()> {
    diagnose_missing_call_like(node, context, diagnostics, "]]")
}

fn diagnose_missing_call_like(
    node: Node,
    context: &DiagnosticContext,
    diagnostics: &mut Vec<Diagnostic>,
    close_token: &str,
) -> anyhow::Result<()> {
    let Some(arguments) = node.child_by_field_name("arguments") else {
        return Ok(());
    };

    diagnose_missing_close(arguments, context, diagnostics, close_token)
}

fn diagnose_missing_binary_operator(
    node: Node,
    context: &DiagnosticContext,
    diagnostics: &mut Vec<Diagnostic>,
) -> anyhow::Result<()> {
    let Some(rhs) = node.child_by_field_name("rhs") else {
        return Ok(());
    };

    if !rhs.is_missing() {
        // Everything is normal
        return Ok(());
    }

    let Some(operator) = node.child_by_field_name("operator") else {
        return Ok(());
    };

    let range = operator.range();

    let text = context.contents.node_slice(&operator)?;
    let message = format!("Invalid binary operator '{text}'. Missing a right hand side.");

    diagnostics.push(new_syntax_diagnostic(message, range, context));

    Ok(())
}

// For namespace operators, the RHS is actually optional in the grammar,
// to help with autocomplete, so we are looking for when this is `None`.
//
// This means that `dplyr::` is actually "valid" R code according to the
// grammar, so this issue won't ever show up in the syntactic path, even
// though its a syntax problem. Instead we expose it from here and use it
// in the semantic path.
pub(crate) fn diagnose_missing_namespace_operator(
    node: Node,
    context: &DiagnosticContext,
    diagnostics: &mut Vec<Diagnostic>,
) -> anyhow::Result<()> {
    let None = node.child_by_field_name("rhs") else {
        // Everything is normal
        return Ok(());
    };

    let Some(operator) = node.child_by_field_name("operator") else {
        return Ok(());
    };

    let range = operator.range();

    let text = context.contents.node_slice(&operator)?;
    let message = format!("Invalid namespace operator '{text}'. Missing a right hand side.");

    diagnostics.push(new_syntax_diagnostic(message, range, context));

    Ok(())
}

// `node` must have required `"open"` and `"close"` fields
fn diagnose_missing_close(
    node: Node,
    context: &DiagnosticContext,
    diagnostics: &mut Vec<Diagnostic>,
    close_token: &str,
) -> anyhow::Result<()> {
    let Some(close) = node.child_by_field_name("close") else {
        return Ok(());
    };

    if !close.is_missing() {
        // Everything is normal
        return Ok(());
    }

    let Some(open) = node.child_by_field_name("open") else {
        return Ok(());
    };

    diagnostics.push(new_missing_close_diagnostic(
        close_token,
        open.range(),
        context,
    ));

    Ok(())
}

fn new_missing_open_diagnostic(
    open_token: &str,
    range: Range,
    context: &DiagnosticContext,
) -> Diagnostic {
    let message = format!("Unmatched closing delimiter. Missing an opening '{open_token}'.");
    new_syntax_diagnostic(message, range, context)
}

fn new_missing_close_diagnostic(
    close_token: &str,
    range: Range,
    context: &DiagnosticContext,
) -> Diagnostic {
    let message = format!("Unmatched opening delimiter. Missing a closing '{close_token}'.");
    new_syntax_diagnostic(message, range, context)
}

fn new_syntax_diagnostic(message: String, range: Range, context: &DiagnosticContext) -> Diagnostic {
    let range = convert_tree_sitter_range_to_lsp_range(context.contents, range);
    Diagnostic::new_simple(range, message)
}

#[cfg(test)]
mod tests {
    use tower_lsp::lsp_types::Diagnostic;
    use tower_lsp::lsp_types::Position;

    use crate::lsp::diagnostics::DiagnosticContext;
    use crate::lsp::diagnostics_syntax::syntax_diagnostics;
    use crate::lsp::documents::Document;

    fn text_diagnostics(text: &str) -> Vec<Diagnostic> {
        let document = Document::new(text, None);
        let context = DiagnosticContext::new(&document.contents);
        let diagnostics = syntax_diagnostics(document.ast.root_node(), &context).unwrap();
        diagnostics
    }

    #[test]
    fn test_syntax_error_truncation() {
        // Coded to truncate at 20 rows
        let newlines = "\n".repeat(20);
        let text = String::from("('") + newlines.as_str() + ")";
        let diagnostics = text_diagnostics(text.as_str());
        let diagnostic = diagnostics.get(0).unwrap();
        insta::assert_snapshot!(diagnostic.message);
        assert_eq!(diagnostic.range.start, Position::new(0, 0));
        assert_eq!(diagnostic.range.end, Position::new(0, 0));
    }

    #[test]
    fn test_unmatched_call_delimiter() {
        let diagnostics = text_diagnostics("match(a, b");
        assert_eq!(diagnostics.len(), 1);
        let diagnostic = diagnostics.get(0).unwrap();
        assert_eq!(diagnostic.range.start, Position::new(0, 5));
        assert_eq!(diagnostic.range.end, Position::new(0, 6));
        insta::assert_snapshot!(diagnostic.message);

        let diagnostics = text_diagnostics("foo[a, b");
        assert_eq!(diagnostics.len(), 1);
        let diagnostic = diagnostics.get(0).unwrap();
        assert_eq!(diagnostic.range.start, Position::new(0, 3));
        assert_eq!(diagnostic.range.end, Position::new(0, 4));
        insta::assert_snapshot!(diagnostic.message);

        let diagnostics = text_diagnostics("foo[[a, b");
        assert_eq!(diagnostics.len(), 1);
        let diagnostic = diagnostics.get(0).unwrap();
        assert_eq!(diagnostic.range.start, Position::new(0, 3));
        assert_eq!(diagnostic.range.end, Position::new(0, 5));
        insta::assert_snapshot!(diagnostic.message);
    }

    #[test]
    fn test_unmatched_call_delimiter_with_trailing_info() {
        // Expect 2 diagnostics
        // - One about unmatched `(`
        // - But the `)` is implied, meaning that between `2` and `identity` there should be a `,`
        //   so we get a diagnostic for that too
        let text = "
match(1, 2

identity(1)
";
        let diagnostics = text_diagnostics(text);
        assert_eq!(diagnostics.len(), 1);

        // Diagnostic highlights the unmatched `(`
        let diagnostic = diagnostics.get(0).unwrap();
        insta::assert_snapshot!(diagnostic.message);
        assert_eq!(diagnostic.range.start, Position::new(1, 5));
        assert_eq!(diagnostic.range.end, Position::new(1, 6));
    }

    #[test]
    fn test_unmatched_braces() {
        let diagnostics = text_diagnostics("{");
        assert_eq!(diagnostics.len(), 1);
        let diagnostic = diagnostics.get(0).unwrap();
        insta::assert_snapshot!(diagnostic.message);

        let diagnostics = text_diagnostics("{ 1 + 2");
        assert_eq!(diagnostics.len(), 1);
        let diagnostic = diagnostics.get(0).unwrap();
        insta::assert_snapshot!(diagnostic.message);

        let diagnostics = text_diagnostics("{}");
        assert!(diagnostics.is_empty());

        let diagnostics = text_diagnostics("{ 1 + 2 }");
        assert!(diagnostics.is_empty());
    }

    #[test]
    fn test_unmatched_parentheses() {
        let diagnostics = text_diagnostics("(");
        assert_eq!(diagnostics.len(), 1);
        let diagnostic = diagnostics.get(0).unwrap();
        insta::assert_snapshot!(diagnostic.message);

        let diagnostics = text_diagnostics("( 1 + 2");
        assert_eq!(diagnostics.len(), 1);
        let diagnostic = diagnostics.get(0).unwrap();
        insta::assert_snapshot!(diagnostic.message);

        let diagnostics = text_diagnostics("()");
        assert!(diagnostics.is_empty());

        let diagnostics = text_diagnostics("( 1 + 2 )");
        assert!(diagnostics.is_empty());
    }

    #[test]
    fn test_error_precision() {
        // The actual error is up to tree-sitter-r's error recovery,
        // but it should always be decent
        let diagnostics = text_diagnostics("sum(1 * 2 + )");
        assert_eq!(diagnostics.len(), 1);
        let diagnostic = diagnostics.get(0).unwrap();
        insta::assert_snapshot!(diagnostic.message);
        assert_eq!(diagnostic.range.start, Position::new(0, 12));
        assert_eq!(diagnostic.range.end, Position::new(0, 13));
    }

    #[test]
    fn test_unmatched_closing_token() {
        let close = vec!["}", ")", "]"];

        for delimiter in close.iter() {
            // i.e. `1 + 1 }`
            let text = format!("1 + 1 {delimiter}");

            let diagnostics = text_diagnostics(text.as_str());
            assert_eq!(diagnostics.len(), 1);

            // Diagnostic highlights the `{delimiter}`
            let diagnostic = diagnostics.get(0).unwrap();
            insta::assert_snapshot!(diagnostic.message);
            assert_eq!(diagnostic.range.start, Position::new(0, 6));
            assert_eq!(diagnostic.range.end, Position::new(0, 7));
        }
    }

    #[test]
    fn test_unmatched_closing_token_precision() {
        // Related to https://github.com/tree-sitter/tree-sitter/issues/3623
        // Should target the `}` specifically, not `+ }`.
        let text = "1 + }";
        let diagnostics = text_diagnostics(text);
        let diagnostic = diagnostics.get(0).unwrap();
        insta::assert_snapshot!(diagnostic.message);
        assert_eq!(diagnostic.range.start, Position::new(0, 4));
        assert_eq!(diagnostic.range.end, Position::new(0, 5));
    }

    #[test]
    fn test_unmatched_binary_operator() {
        // The actual error is up to tree-sitter-r's error recovery,
        // but it should always be decent
        let text = "
{
 1 +
}";

        let diagnostics = text_diagnostics(text);
        assert_eq!(diagnostics.len(), 1);

        let diagnostic = diagnostics.get(0).unwrap();
        insta::assert_snapshot!(diagnostic.message);
        assert_eq!(diagnostic.range.start, Position::new(3, 0));
        assert_eq!(diagnostic.range.end, Position::new(3, 1));
    }

    #[test]
    fn test_unmatched_function_parameters_parentheses() {
        // Exact set of diagnostics are up to tree-sitter-r's error recovery,
        // but they should be decent at pointing you to the right place
        let text = "
function(x {
  1 + 1
}";

        let diagnostics = text_diagnostics(text);
        assert_eq!(diagnostics.len(), 2);

        let diagnostic = diagnostics.get(0).unwrap();
        insta::assert_snapshot!(diagnostic.message);
        assert_eq!(diagnostic.range.start, Position::new(1, 11));
        assert_eq!(diagnostic.range.end, Position::new(1, 12));

        let diagnostic = diagnostics.get(1).unwrap();
        insta::assert_snapshot!(diagnostic.message);
        assert_eq!(diagnostic.range.start, Position::new(3, 0));
        assert_eq!(diagnostic.range.end, Position::new(3, 1));
    }
}
