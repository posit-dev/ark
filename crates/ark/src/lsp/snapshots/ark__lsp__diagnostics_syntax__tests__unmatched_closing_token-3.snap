---
source: crates/ark/src/lsp/diagnostics_syntax.rs
expression: diagnostic.message
---
Unmatched closing delimiter. Missing an opening '['.
